File: smite/src/main/java/hirez/smite/object/God.java
package hirez.smite.object;

import com.fasterxml.jackson.annotation.JsonProperty;
import hirez.api.object.adapters.TextToBoolean;
import hirez.api.object.interfaces.ReturnedMessage;
import lombok.Data;

import java.util.List;

@Data
public class God implements ReturnedMessage {
    @JsonProperty("Ability_1")
    private final Ability ability1;
    @JsonProperty("Ability_2")
    private final Ability ability2;
    @JsonProperty("Ability_3")
    private final Ability ability3;
    @JsonProperty("Ability_4")
    private final Ability ability4;
    @JsonProperty("Ability_5")
    private final Ability ability5;

    @JsonProperty("Ability1")
    private final String ability1Name;
    @JsonProperty("Ability2")
    private final String ability2Name;
    @JsonProperty("Ability3")
    private final String ability3Name;
    @JsonProperty("Ability4")
    private final String ability4Name;
    @JsonProperty("Ability5")
    private final String ability5Name;

    @JsonProperty("abilityDescription1")
    private final ItemDescription abilityDescription1;
    @JsonProperty("abilityDescription2")
    private final ItemDescription abilityDescription2;
    @JsonProperty("abilityDescription3")
    private final ItemDescription abilityDescription3;
    @JsonProperty("abilityDescription4")
    private final ItemDescription abilityDescription4;
    @JsonProperty("abilityDescription5")
    private final ItemDescription abilityDescription5;

    @JsonProperty("AbilityId1")
    private final long ability1Id;
    @JsonProperty("AbilityId2")
    private final long ability2Id;
    @JsonProperty("AbilityId3")
    private final long ability3Id;
    @JsonProperty("AbilityId4")
    private final long ability4Id;
    @JsonProperty("AbilityId5")
    private final long ability5Id;

    @JsonProperty("AttackSpeed")
    private final double attackSpeed;
    @JsonProperty("AttackSpeedPerLevel")
    private final double attackSpeedPerLevel;
    @JsonProperty("basicAttack")
    private final ItemDescription basicAttack;
    @JsonProperty("Cons")
    private final String cons;

    @JsonProperty("godAbility1_URL")
    private final String ability1URL;
    @JsonProperty("godAbility2_URL")
    private final String ability2URL;
    @JsonProperty("godAbility3_URL")
    private final String ability3URL;
    @JsonProperty("godAbility4_URL")
    private final String ability4URL;
    @JsonProperty("godAbility5_URL")
    private final String ability5URL;

    @JsonProperty("godCard_URL")
    private final String cardURL;
    @JsonProperty("godIcon_URL")
    private final String iconURL;

    @JsonProperty("HP5PerLevel")
    private final double hp5PerLevel;
    @JsonProperty("Health")
    private final int health;
    @JsonProperty("HealthPerFive")
    private final int healthPerFive;
    @JsonProperty("HealthPerLevel")
    private final int healthPerLevel;
    private final long id;
    @TextToBoolean
    @JsonProperty("latestGod")
    private final boolean latestGod;
    @JsonProperty("Lore")
    private final String lore;
    @JsonProperty("MP5PerLevel")
    private final double mp5PerLevel;
    @JsonProperty("MagicProtection")
    private final int magicProtection;
    @JsonProperty("MagicProtectionPerLevel")
    private final double magicProtectionPerLevel;
    @JsonProperty("MagicalPower")
    private final int magicalPower;
    @JsonProperty("MagicalPowerPerLevel")
    private final int magicalPowerPerLevel;
    @JsonProperty("Mana")
    private final int mana;
    @JsonProperty("ManaPerFive")
    private final double manaPerFive;
    @JsonProperty("ManaPerLevel")
    private final int manaPerLevel;
    @JsonProperty("Name")
    private final String name;
    @TextToBoolean
    @JsonProperty("OnFreeRotation")
    private final boolean onFreeRotation;
    @JsonProperty("Pantheon")
    private final String pantheon;
    @JsonProperty("PhysicalPower")
    private final int physicalPower;
    @JsonProperty("PhysicalPowerPerLevel")
    private final int physicalPowerPerLevel;
    @JsonProperty("PhysicalProtection")
    private final int physicalProtection;
    @JsonProperty("PhysicalProtectionPerLevel")
    private final int physicalProtectionPerLevel;
    @JsonProperty("Pros")
    private final String pros;
    @JsonProperty("ret_msg")
    private final String returnedMessage;
    @JsonProperty("Roles")
    private final String roles;
    @JsonProperty("Speed")
    private final int speed;
    @JsonProperty("Title")
    private final String title;
    @JsonProperty("Type")
    private final String type;

    @Data
    public static class Description {
        private final String cooldown;
        private final String cost;
        private final String description;
        @JsonProperty("menuitems")
        private final List<DescriptionValue> menuItems;
        @JsonProperty("rankitems")
        private final List<DescriptionValue> rankItems;
        private final String secondaryDescription;
    }

    @Data
    public static class ItemDescription {
        @JsonProperty("itemDescription")
        private final Description itemDescription;
    }

    @Data
    public static class Ability {
        @JsonProperty("Description")
        private final ItemDescription description;
        @JsonProperty("Id")
        private final long id;
        @JsonProperty("Summary")
        private final String name;
        @JsonProperty("URL")
        private final String url;
    }
}

File: smite/src/main/kotlin/hirez/smite/json/god.kt
package hirez.smite.json

import com.google.gson.annotations.JsonAdapter
import com.google.gson.annotations.SerializedName
import hirez.Language
import hirez.api.Queue
import hirez.json.ReturnMessage
import hirez.json.adapters.BooleanTextDeserializer
import hirez.smite.SmiteGame

/**
 *
 * @author Damian Staszewski [damian@stachuofficial.tv]
 * @version %I%, %G%
 * @since 1.0
 */
data class God(
			@SerializedName("Ability_1")
			val ability1: God.Ability,
			@SerializedName("Ability_2")
			val ability2: God.Ability,
			@SerializedName("Ability_3")
			val ability3: God.Ability,
			@SerializedName("Ability_4")
			val ability4: God.Ability,
			@SerializedName("Ability_5")
			val ability5: God.Ability,
			@SerializedName("ret_msg")
			override val returnMessage: String?,
			@SerializedName("AttackSpeed")
			val attackSpeed: Float,
			@SerializedName("AttackSpeedPerLevel")
			val attackSpeedPerLevel: Float,
			@SerializedName("Cons")
			val cons: String,
			@SerializedName("HP5PerLevel")
			val hP5PerLevel: Float,
			@SerializedName("Health")
			val health: Float,
			@SerializedName("HealthPerFive")
			val healthPerFive: Float,
			@SerializedName("HealthPerLevel")
			val healthPerLevel: Float,
			@SerializedName("Lore")
			val lore: String,
			@SerializedName("MP5PerLevel")
			val mP5PerLevel: Float,
			@SerializedName("MagicProtection")
			val magicProtection: Float,
			@SerializedName("MagicProtectionPerLevel")
			val magicProtectionPerLevel: Float,
			@SerializedName("MagicalPower")
			val magicalPower: Float,
			@SerializedName("MagicalPowerPerLevel")
			val magicalPowerPerLevel: Float,
			@SerializedName("Mana")
			val mana: Float,
			@SerializedName("ManaPerFive")
			val manaPerFive: Float,
			@SerializedName("ManaPerLevel")
			val manaPerLevel: Int,
			@SerializedName("Name")
			val name: String,
			@SerializedName("OnFreeRotation")
			@JsonAdapter(BooleanTextDeserializer::class)
			val freeRotation: Boolean,
			@SerializedName("Pantheon")
			val pantheon: String,
			@SerializedName("PhysicalPower")
			val physicalPower: Float,
			@SerializedName("PhysicalPowerPerLevel")
			val physicalPowerPerLevel: Float,
			@SerializedName("PhysicalProtection")
			val physicalProtection: Float,
			@SerializedName("PhysicalProtectionPerLevel")
			val physicalProtectionPerLevel: Double,
			@SerializedName("Pros")
			val pros: String,
			@SerializedName("Roles")
			val roles: String,
			@SerializedName("Speed")
			val speed: Float,
			@SerializedName("Title")
			val title: String,
			@SerializedName("Type")
			val type: String,
			val basicAttack: BaseDescription,
			@SerializedName("godCard_URL")
			val cardUrl: String,
			@SerializedName("godIcon_URL")
			val iconUrl: String,
			val id: Long,
			@JsonAdapter(BooleanTextDeserializer::class)
			val latestGod: Boolean
) : ReturnMessage {
	val abilityDescription1
		get() = ability1.description
	val abilityDescription2
		get() = ability2.description
	val abilityDescription3
		get() = ability3.description
	val abilityDescription4
		get() = ability4.description
	val abilityDescription5
		get() = ability5.description
	
	fun getLeaderboard(queue: Queue) =
				SmiteGame.getGodLeaderboard(id, queue)
	
	fun getSkins(language: Language) =
				SmiteGame.getGodSkins(id, language)
	
	val skins
		get() = SmiteGame.getGodSkins(id)
	
	fun getRecommendedItems(language: Language) =
				SmiteGame.getGodRecommendedItems(id, language)
	
	val recommendedItems
		get() = SmiteGame.getGodRecommendedItems(id)
	
	data class Ability(
				@SerializedName("Description")
				val description: BaseDescription,
				@SerializedName("Id")
				val id: Long,
				@SerializedName("Summary")
				val name: String,
				@SerializedName("URL")
				val abilityImage: String
	)
}

data class BaseDescription(
			val itemDescription: ItemedDescription
) {
	data class ItemedDescription(
				val cooldown: String,
				val cost: String,
				val description: String,
				val menuitems: List<Item>,
				val rankitems: List<Item>,
				val secondaryDescription: String?
	)
	
	data class Item(
				val description: String,
				val value: String
	)
}

data class GodSkin(
			@SerializedName("godIcon_URL")
			val iconUrl: String,
			@SerializedName("godSkin_URL")
			val skinUrl: String,
			@SerializedName("god_id")
			val godId: Long,
			@SerializedName("god_name")
			val godName: String,
			val obtainability: String,
			@SerializedName("price_favor")
			val priceFavor: Int,
			@SerializedName("price_gems")
			val priceGems: Int,
			@SerializedName("skin_id1")
			val Id1: Long,
			@SerializedName("skin_id2")
			val Id2: Long,
			@SerializedName("skin_name")
			val name: String,
			override val returnMessage: String?
) : ReturnMessage

data class GodLeaderboard(
			@SerializedName("ret_msg")
			override val returnMessage: String?,
			val godId: Long,
			val losses: Int,
			val playerId: Long,
			val playerName: String,
			val playerRanking: Float,
			val rank: Int,
			val wins: Int
) : ReturnMessage